GliaWidgets/SecureConversations/SecureConversations.MessagesWithUnreadCountLoader.swift

```diff
@@ -17,64 +17,91 @@ extension SecureConversations {
         var environment: Environment
 
         func loadMessagesWithUnreadCount(callback: @escaping (Result<MessagesWithUnreadCount, Error>) -> Void) {
-            typealias ReactiveSwift = CoreSdkClient.ReactiveSwift
-            let unreadCountTimeoutProducer = ReactiveSwift.SignalProducer<Void, Never>(value: ())
-            let unreadCountRequestSource = ReactiveSwift.Signal<Result<Int, Error>, Never>.pipe()
-            let unreadCountTimeoutSource = ReactiveSwift.Signal<Result<Int, Error>, Never>.pipe()
-            // In case if `unreadCount` will never be delivered,
-            // (for example if sockets are disconnected, because
-            // `unreadCount` is based on sockets)
-            // we will still receive callback by timeout.
-            let unreadCountRequestWithTimeoutSource = ReactiveSwift.Signal.merge(
-                unreadCountRequestSource.output,
-                unreadCountTimeoutSource.output
-            ).take(first: 1)
+            let stateQueue = DispatchQueue(label: "MessagesWithUnreadCountLoader.state")
+            var didCallback = false
+            var messagesResult: Result<[ChatMessage], Error>?
+            var unreadResult: Result<Int, Error>?
 
-            let messagesSource = ReactiveSwift.Signal<Result<[ChatMessage], CoreSdkClient.SalemoveError>, Never>.pipe()
-
-            let combined = ReactiveSwift.Signal.zip(
-                messagesSource.output,
-                unreadCountRequestWithTimeoutSource
-            )
-
-            unreadCountTimeoutProducer
-                .delay(Self.unreadCountFallbackTimeoutSeconds, on: environment.scheduler)
-                .startWithCompleted {
-                    unreadCountTimeoutSource.input.send(value: .failure(TimeoutError()))
+            func finishIfReady() {
+                stateQueue.async {
+                    guard !didCallback,
+                          let m = messagesResult,
+                          let u = unreadResult
```

Comment:
Can it be done like this?
```suggestion
                          let messagesResult,
                          let unreadResult
```

Reply:
If it is okay with you, I'd keep it because a PR for fetchChatHistory, that is already ready, removes this workaround logic. `MessagesWithUnreadCountLoader` is much simpler after that. with 2 concurrent async tasks running, not one with async and other with callback.